File: libs/htmlcanvas/src/main/kotlin/mono/html/canvas/canvas/GridCanvasViewController.kt
package mono.html.canvas.canvas

import mono.lifecycle.LifecycleOwner
import mono.livedata.LiveData
import org.w3c.dom.HTMLCanvasElement
import org.w3c.dom.Path2D

internal class GridCanvasViewController(
    lifecycleOwner: LifecycleOwner,
    canvas: HTMLCanvasElement,
    drawingInfoLiveData: LiveData<DrawingInfoController.DrawingInfo>
) : BaseCanvasViewController(canvas) {

    init {
        drawingInfoLiveData.observe(lifecycleOwner, listener = ::setDrawingInfo)
        drawingInfoLiveData.observe(lifecycleOwner, 0) { draw() }
    }

    override fun drawInternal() {
        context.strokeStyle = "#d9d9d9"
        context.lineWidth = 0.25
        context.stroke(createGridPath())

        context.strokeStyle = "#BBBBBB"
        context.lineWidth = 1.0
        val zeroLines = Path2D().apply {
            addHLine(
                0.0,
                drawingInfo.toYPx(0.0),
                drawingInfo.canvasSizePx.width.toDouble()
            )
            addVLine(
                drawingInfo.toXPx(0.0),
                0.0,
                drawingInfo.canvasSizePx.height.toDouble()
            )
        }
        context.stroke(zeroLines)
    }

    private fun createGridPath(): Path2D = Path2D().apply {
        val zeroX = drawingInfo.toXPx(drawingInfo.boardColumnRange.first.toDouble() - 1.0)
        val maxX = drawingInfo.toXPx(drawingInfo.boardColumnRange.last.toDouble() + 1.0)
        val zeroY = drawingInfo.toYPx(drawingInfo.boardRowRange.first.toDouble() - 1.0)
        val maxY = drawingInfo.toYPx(drawingInfo.boardRowRange.last.toDouble() + 1.0)

        for (row in drawingInfo.boardRowRange) {
            val y = drawingInfo.toYPx(row.toDouble())
            addHLine(zeroX, y, maxX - zeroX)
        }

        for (col in drawingInfo.boardColumnRange) {
            val x = drawingInfo.toXPx(col.toDouble())
            addVLine(x, zeroY, maxY - zeroY)
        }
    }
}

File: monosketch-svelte/src/lib/mono/workspace/canvas/grid-canvas-view-controller.ts
import type { AppUiStateManager } from "$mono/ui-state-manager/app-ui-state-manager";
import { BaseCanvasViewController } from '$mono/workspace/canvas/base-canvas-controller';
import { ThemeColors } from '$mono/ui-state-manager/states';

/**
 * A class for managing the grid canvas.
 */
export class GridCanvasViewController extends BaseCanvasViewController {
    constructor(
        canvas: HTMLCanvasElement,
        private appUiStateManager: AppUiStateManager,
    ) {
        super(canvas);
    }

    protected drawInternal = () => {
        const context = this.context;
        context.strokeStyle = this.appUiStateManager.getThemedColorCode(ThemeColors.GridLine);
        context.lineWidth = 0.25;
        context.stroke(this.createGridPath());
    };

    private createGridPath = (): Path2D => {
        const drawingInfo = this.drawingInfo;
        const zeroX = drawingInfo.toXPx(drawingInfo.boardColumnRange.start - 1.0);
        const maxX = drawingInfo.toXPx(drawingInfo.boardColumnRange.endExclusive + 1.0);
        const zeroY = drawingInfo.toYPx(drawingInfo.boardRowRange.start - 1.0);
        const maxY = drawingInfo.toYPx(drawingInfo.boardRowRange.endExclusive + 1.0);

        const path = new Path2D();

        for (const row of drawingInfo.boardRowRange) {
            const yPx = drawingInfo.toYPx(row);
            this.addHLine(path, zeroX, yPx, maxX - zeroX);
        }

        for (const column of drawingInfo.boardColumnRange) {
            const xPx = drawingInfo.toXPx(column);
            this.addVLine(path, xPx, zeroY, maxY - zeroY);
        }

        return path;
    };
}
